ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -1303,3 +1337,233 @@ bool attribute_is_exportable(struct pkcs11_attribute_head *req_attr,
 
 	return true;
 }
+
+static bool attr_is_modifiable_any_key(struct pkcs11_attribute_head *attr)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_ID:
+	case PKCS11_CKA_START_DATE:
+	case PKCS11_CKA_END_DATE:
+	case PKCS11_CKA_DERIVE:
+		return true;
+	default:
+		return false;
+	}
+}
+
+static bool attr_is_modifiable_secret_key(struct pkcs11_attribute_head *attr,
+					  struct pkcs11_session *session,
+					  struct pkcs11_object *obj)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_ENCRYPT:
+	case PKCS11_CKA_DECRYPT:
+	case PKCS11_CKA_SIGN:
+	case PKCS11_CKA_VERIFY:
+	case PKCS11_CKA_WRAP:
+	case PKCS11_CKA_UNWRAP:
+		return true;
+	/* Can't be modified once set to CK_FALSE - 12 in Table 10 */
+	case PKCS11_CKA_EXTRACTABLE:
+		return get_bool(obj->attributes, PKCS11_CKA_EXTRACTABLE);
+	/* Can't be modified once set to CK_TRUE - 11 in Table 10 */
+	case PKCS11_CKA_SENSITIVE:
+		return !get_bool(obj->attributes, PKCS11_CKA_SENSITIVE);
+	case PKCS11_CKA_WRAP_WITH_TRUSTED:
+		return !get_bool(obj->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED);
+	/* Change in CKS_TRUSTED can only be done by SO */
```

Comment:
CKA_TRUSTED

I think this is applicable to all objects. Maybe move to `attribute_is_modifiable()` generic sequence.

Reply:
It is applicable to only key objects. I can move it to any_key_modifiable but would prefer to keep it here so that attributes in each of this function match with the tables in PKCS#11 specification marked 8,10, 11 or 12. It is easier to remember and trace that way. Let me know what you think.

Reply:
@ruchi393 please note the typo fix as mentioned by @etienne-lms

Reply:
Thanks , I missed the typo fix. Will correct that.